models/percona_sso_model_helpers.go

```diff
@@ -17,32 +17,128 @@
 package models
 
 import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"sync"
+	"time"
+
 	"github.com/pkg/errors"
 	"gopkg.in/reform.v1"
 )
 
+var perconaSSOMtx sync.Mutex
+
 // GetPerconaSSODetails returns PerconaSSODetails if there are any, error otherwise.
-func GetPerconaSSODetails(q *reform.Querier) (*PerconaSSODetails, error) {
-	ssoDetails, err := q.SelectOneFrom(PerconaSSODetailsView, "")
+// Access token is automatically refreshed if it is expired.
+// Get, check eventually refresh done in one tx.
+func GetPerconaSSODetails(ctx context.Context, q *reform.Querier) (*PerconaSSODetails, error) {
+	perconaSSOMtx.Lock()
+	defer perconaSSOMtx.Unlock()
+
+	ssoDetails, err := q.SelectOneFrom(PerconaSSODetailsTable, "")
 	if err != nil {
 		return nil, errors.Wrap(err, "failed to get Percona SSO Details")
 	}
-	return ssoDetails.(*PerconaSSODetails), nil
+
+	details := ssoDetails.(*PerconaSSODetails)
+	if details.isAccessTokenExpired() {
+		refreshedToken, err := details.refreshAndGetAccessToken(ctx, q)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to insert Percona SSO Details")
+		}
+		details.AccessToken = refreshedToken
+	}
+
+	return details, nil
+}
+
+// GetPerconaSSOAccessToken returns PerconaSSOAccessToken if there is GetPerconaSSODetails, error otherwise.
+func GetPerconaSSOAccessToken(ctx context.Context, q *reform.Querier) (*PerconaSSOAccessToken, error) {
+	ssoDetails, err := GetPerconaSSODetails(ctx, q)
+	if err != nil {
+		return nil, err
+	}
+
+	return ssoDetails.AccessToken, nil
+}
+
+func (sso *PerconaSSODetails) refreshAndGetAccessToken(ctx context.Context, q *reform.Querier) (*PerconaSSOAccessToken, error) {
+	url := fmt.Sprintf("%s%s", sso.IssuerURL, sso.Scope)
```

Comment:
1) this is not enough, please check https://confluence.percona.com/display/PMM/Single+Sign-On+-+PMM+Server+integration+with+Portal+and+Okta#SingleSignOnPMMServerintegrationwithPortalandOkta-ObtainandUseAccessToken

```
curl -X POST '<ISSUER URL>/token?grant_type=client_credentials&scope=<portal_auth_scope>' \
--header 'Content-Type: application/x-www-form-urlencoded' \
--header 'accept: application/json' \
--header 'Authorization: Basic base64(<portal_auth_client_id:portal_auth_client_secret>)'
```

`/token` and `grant_type=client_credentials` are missed.

2) use encoded url params